syntax/options.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2023 The Bazel Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package syntax
+
+import _ "unsafe" // for linkname
+
+// FileOptions specifies various per-file options that affect static
+// aspects of an individual file such as parsing, name resolution, and
+// code generation. (Options that affect global dynamics are typically
+// controlled through [starlark.Thread].)
+//
+// The zero value of FileOptions is the default behavior.
+//
+// Many functions in this package come in two versions: the legacy
+// standalone function (such as [Parse]) uses [LegacyFileOptions],
+// whereas the more recent method (such as [Options.Parse]) honors the
+// provided options. The second form is preferred. In other packages,
+// the modern version is a standalone function with a leading
+// FileOptions parameter and the name suffix "Options", such as
+// [starlark.ExecFileOptions].
+type FileOptions struct {
+	// resolver
+	Set               bool // allow references to the 'set' built-in function
+	While             bool // allow 'while' statements
+	TopLevelControl   bool // allow if/for/while statements at top-level
+	GlobalReassign    bool // allow reassignment to top-level names
+	LoadBindsGlobally bool // load creates global not file-local bindings (deprecated)
+
+	// compiler
+	Recursion bool // disable recursion check for functions in this file
+}
+
+// TODO(adonovan): provide a canonical flag parser for FileOptions.
+// (And use it in the testdata "options:" strings.)
+
+// LegacyFileOptions returns a new FileOptions containing the current
+// values of the resolver package's legacy global variables such as
+// [resolve.AllowRecursion], etc. These may be connected to
+// command-line flags.
+func LegacyFileOptions() *FileOptions {
+	return &FileOptions{
+		Set:               resolverAllowSet,
+		While:             resolverAllowGlobalReassign,
+		TopLevelControl:   resolverAllowGlobalReassign,
```

Comment:
I decided to split the flag into finer features in this CL.   The -recursion flag is required to use `while` statements anywhere because potentially unbounded loops are expressively equivalent to recursive calls with arbitrary arguments. 

As for -globalreassign, the flag is required only to use 'while' (or if or for) at top-level. I think the reason for this was so that Bazel could enforce the invariant that every top-level variable had exactly one assignment, also at top-level. Allowing re-declaration of globals (e.g. x=1; x=2) obviously breaks this invariant. More subtly, so does a loop `for x in seq: pass`: although there is only one static assignment, it is executed dynamically many times, including zero. And so too does `if cond: x = 1`, because it can cause a variable to be initialized zero times. (An if statement is an at-most-once loop!) Obviously `while` is like `for`. Personally I don't think the concern is important, and this flag is too hard to explain. @laurentlb may remember more.